File: typescript/hexadecimal_to_decimal_converter/hexadecimal_to_decimal_converter_25.ts
/**
 * This TypeScript program is designed to convert hexadecimal numbers to their decimal equivalents.
 * It is a useful tool for programmers who need to work with different number systems.
 * The program includes a function to generate random numbers, which can be used for various purposes.
 * The main function of the program is to take a hexadecimal string as input and return its decimal value.
 * The program also includes extensive comments to help understand the flow and logic of the code.
 */

// Function to generate a random number
function generateRandomNumber(): number {
    // Initialize the seed for the random number generator
    let seed = 1337;
    let randomNumber = Math.sin(seed++) * 10000;
    return randomNumber - Math.floor(randomNumber);
}

// Function to convert a single hexadecimal digit to its decimal equivalent
function hexDigitToDecimal(hexDigit: string): number {
    const hexMap: { [key: string]: number } = {
        '0': 0, '1': 1, '2': 2, '3': 3, '4': 4, '5': 5, '6': 6, '7': 7,
        '8': 8, '9': 9, 'A': 10, 'B': 11, 'C': 12, 'D': 13, 'E': 14, 'F': 15
    };
    return hexMap[hexDigit.toUpperCase()];
}

// Function to convert a hexadecimal string to its decimal equivalent
function hexToDecimal(hexString: string): number {
    let decimalValue = 0;
    let power = 0;
    for (let i = hexString.length - 1; i >= 0; i--) {
        let currentDigit = hexString[i];
        decimalValue += hexDigitToDecimal(currentDigit) * Math.pow(16, power);
        power++;
    }
    return decimalValue;
}

// Main function to demonstrate the conversion
function main() {
    // Example hexadecimal strings
    let hex1 = "1A3F";
    let hex2 = "FF";
    let hex3 = "123ABC";

    // Convert and display the decimal equivalents
    console.log(`The decimal equivalent of ${hex1} is ${hexToDecimal(hex1)}`);
    console.log(`The decimal equivalent of ${hex2} is ${hexToDecimal(hex2)}`);
    console.log(`The decimal equivalent of ${hex3} is ${hexToDecimal(hex3)}`);

    // Generate and display a random number
    console.log(`A random number: ${generateRandomNumber()}`);
}

// Call the main function to run the program
main();

/***
 */

File: swift/hexadecimal_to_decimal_converter/hexadecimal_to_decimal_converter_25.swift
//
// This program is a hexadecimal to decimal converter written in Swift.
// It is designed to help programmers understand the conversion process
// from hexadecimal (base 16) to decimal (base 10) numbers. The program
// will prompt the user to enter a hexadecimal number and then output
// the corresponding decimal value. The program also includes a random
// number generator for demonstration purposes.
//

import Foundation

// Function to convert a single hexadecimal character to its decimal value
func hexCharToDecimal(_ char: Character) -> Int {
    let hexDigits = "0123456789ABCDEF"
    if let index = hexDigits.firstIndex(of: char) {
        return hexDigits.distance(from: hexDigits.startIndex, to: index)
    }
    return -1 // Return -1 for invalid characters
}

// Function to convert a hexadecimal string to a decimal integer
func hexToDecimal(_ hex: String) -> Int {
    var decimalValue = 0
    var power = 0
    for char in hex.uppercased().reversed() {
        let decimalDigit = hexCharToDecimal(char)
        if decimalDigit == -1 {
            print("Invalid hexadecimal character: \(char)")
            return -1
        }
        decimalValue += decimalDigit * Int(pow(16.0, Double(power)))
        power += 1
    }
    return decimalValue
}

// Function to generate a random number
func generateRandomNumber() -> Int {
    let seed = 1337
    srand48(seed)
    return Int(drand48() * 100)
}

// Main function to execute the program
func main() {
    // Prompt the user to enter a hexadecimal number
    print("Enter a hexadecimal number:")
    if let input = readLine() {
        // Convert the input hexadecimal number to decimal
        let decimalValue = hexToDecimal(input)
        if decimalValue != -1 {
            print("The decimal value of \(input) is \(decimalValue).")
        }
    }
    
    // Generate and print a random number
    let randomNumber = generateRandomNumber()
    print("Random number: \(randomNumber)")
}

// Call the main function to start the program
main()

